File: python/code_07.py
import numpy as np
from time import perf_counter as pfc

inp = open('./data/dt_07.txt').read().split('\n')
inp = np.array(inp[0].split(','), dtype=int)

### MATHEMATICAL SOLUTION
start = pfc()
#part 1
ans1 = np.sum(np.abs(inp - np.median(inp)))
print(f"Answer 1: {ans1}")
#part 2
val  = np.abs(inp - np.floor(np.mean(inp)))
ans2 = np.sum(val * (val + 1)/2)
print(f"Answer 2: {ans2}")
print(pfc()-start)


## Array of unique crabs' positions
inp_t = np.array(list(set(inp)), dtype = int)
## Count haw many crabs have same position
nn = np.array([np.sum(inp == x) for x in inp_t])

##list of potential positions crabs can move to
potential_pos = np.arange(1, max(inp_t)) 
potential_pos = np.repeat(potential_pos, len(inp_t))

inp_t = np.hstack((inp_t, ) * max(potential_pos))
nn =  np.hstack((nn, ) * max(potential_pos))

## PART 1
start = pfc()
d = np.abs(inp_t - potential_pos) * nn
options = np.array([sum(d[potential_pos==x]) for x in range(1, max(potential_pos))], dtype = np.int64)
print(f"Answer 1: {np.min(options)}")
print(pfc()-start)

## PART 2
start = pfc()
d = np.abs(inp_t - potential_pos)
dd = (d * (d + 1)/2) * nn
options = np.array([sum(dd[potential_pos==x]) for x in range(1, max(potential_pos))], dtype = np.int64)
print(f"Answer 2: {np.min(options)}")
print(pfc()-start)

File: js/code_07.js
const { match } = require('assert')
const fs = require('fs')
const input = fs.readFileSync('./data/dt_07.txt', 'utf8').split(',').map(Number)

/* console
url https://adventofcode.com/2021/day/7/input
const input = document.querySelector("body > pre").innerText.split(',').map(Number)  
*/

const fuel_expenses = (from, to) => {
    const dist = Math.abs(from - to)
    return (dist * (dist + 1) / 2)
}

let min_fuel_expenses1 = Number.MAX_SAFE_INTEGER
let min_fuel_expenses2 = Number.MAX_SAFE_INTEGER

for (let pos1 = 1; pos1 <= Math.max(...input); pos1++){
    let total_fuel_expenses1 = 0
    let total_fuel_expenses2 = 0
    for (let pos2 of input) {
        total_fuel_expenses1 += Math.abs(pos1 - pos2)
        total_fuel_expenses2 += fuel_expenses(pos1, pos2)
    }
    if (total_fuel_expenses1 < min_fuel_expenses1) {
        min_fuel_expenses1 = total_fuel_expenses1;
    }
    
    if (total_fuel_expenses2 < min_fuel_expenses2) {
        min_fuel_expenses2 = total_fuel_expenses2;
    }
}
console.log('PART 1: ' + min_fuel_expenses1)
console.log('PART 2: ' + min_fuel_expenses2)

// MATHEMATICAL SOLUTION
const part1 = (() => {
    const sorted = [...input].sort((a, b) => a - b);
    const median = sorted[Math.floor(sorted.length / 2)]
    return sorted.reduce((sum, n) => sum + Math.abs(n - median), 0);
})()

const part2 = (() => {
    const mean = Math.floor(input.reduce((sum, n) => sum + n) / input.length);
    const potential_mean = [mean - 1, mean, mean + 1];
    const calcDistance = n => (n * (n + 1)) / 2;
    const total_fuel_expenses = potential_mean.map(m => {
      return input.reduce((sum, x) => sum + calcDistance(Math.abs(x - m)), 0)
    })
    return Math.min(...total_fuel_expenses);
})()
console.log(part1);
console.log(part2);